app/models/components/course/achievements_ability_component.rb

```diff
@@ -3,38 +3,34 @@ module Course::AchievementsAbilityComponent
   include AbilityHost::Component
 
   def define_permissions
-    if user
+    if course_user&.student?
       allow_student_show_achievements
       allow_students_with_achievement_show_badges
-      allow_staff_manage_achievements
-      do_not_allow_staff_award_automatically_awarded_achievements
     end
 
+    allow_staff_manage_achievements if course_user&.staff?
+    do_not_allow_staff_award_automatically_awarded_achievements
+
     super
   end
 
   private
 
   def allow_student_show_achievements
-    can :read, Course::Achievement,
-        course_all_course_users_hash.reverse_merge(published_achievement_hash)
+    can :read, Course::Achievement, course_id: course.id, published: true
+  end
+
+  def allow_students_with_achievement_show_badges
+    can :display_badge, Course::Achievement, course_user_hash
```

Comment:
should we replace `course_user_hash`?

Reply:
Okay, I think that way would be more readable.